Main/Presenter/mainpresenter.cpp

```diff
@@ -84,12 +105,109 @@ void MainPresenter::saveActiveProject() {
 }
 
 int MainPresenter::getMasterPitch() {
-    return activeProject->transport->getMasterPitch();
+    return static_cast<int>(activeProject->transport->masterPitch->get() + 0.5f);
 }
 
 void MainPresenter::setMasterPitch(int pitch, bool isFinal) {
-    activeProject->transport->setMasterPitch(pitch);
+    activeProject->transport->masterPitch->set(static_cast<float>(pitch), isFinal);
     if (isFinal) {
         emit masterPitchChanged(pitch);
     }
 }
+
+// The patch logic below assumes that only one operation
+// will happen at once. If two separate model items
+// initiate patch builds, they will be grouped together,
+// and possibly malformed in some cases, especially when
+// it comes to undo/redo.
+
+void MainPresenter::initializeNewPatchIfNeeded() {
+    if (isPatchInProgress) {
+        return;
+    }
+
+    isPatchInProgress = true;
+
+    historyPointer++;
+
+    if (historyPointer != projectHistory.length()) {
+        for (int i = projectHistory.length() - 1; i >= historyPointer; i--) {
+            projectHistory[i]->~Patch();
+            projectHistory.pop_back();
+        }
+    }
+
+    projectHistory.append(
+        new Patch(this, activeProject, projectFiles[activeProjectIndex]->document)
+    );
+}
+
+void MainPresenter::patchAdd(QString path, rapidjson::Value& value) {
+    initializeNewPatchIfNeeded();
+    Patch& patch = *projectHistory[historyPointer];
+    Value copiedValue(value, projectFiles[activeProjectIndex]->document.GetAllocator());
+    patch.patchAdd("/" + path, copiedValue);
+}
+
+void MainPresenter::patchRemove(QString path) {
+    initializeNewPatchIfNeeded();
+    Patch& patch = *projectHistory[historyPointer];
+    patch.patchRemove("/" + path);
+}
+
+void MainPresenter::patchReplace(QString path, rapidjson::Value& value) {
+    initializeNewPatchIfNeeded();
+    Patch& patch = *projectHistory[historyPointer];
+    Value copiedValue(value, projectFiles[activeProjectIndex]->document.GetAllocator());
+    patch.patchReplace("/" + path, copiedValue);
+}
+
+void MainPresenter::patchCopy(QString from, QString path) {
+    initializeNewPatchIfNeeded();
+    Patch& patch = *projectHistory[historyPointer];
+    patch.patchCopy("/" + from, "/" + path);
+}
+
+void MainPresenter::patchMove(QString from, QString path) {
+    initializeNewPatchIfNeeded();
+    Patch& patch = *projectHistory[historyPointer];
+    patch.patchMove("/" + from, "/" + path);
+}
+
+void MainPresenter::sendPatch() {
+    if (!isPatchInProgress) {
+        throw "sendPatch() was called, but there was nothing to send.";
+    }
+    Patch& patch = *projectHistory[historyPointer];
+    patch.apply();
+    engine->sendPatchList(patch.getPatch());
```

Comment:
The engine is not necessarily active for this project or at all: maybe check for that first?

Same with the other `engine->...()` calls later in this file

Reply:
What do you mean by "active"? I haven't added in project switching, but I also don't know what messaging I'd use for that (I'm not reading anything from the engine - if it's running, it's assumed to be active, which is simplistic I know)

Reply:
Multiple projects may be open at the same time, but only one engine is ever running. Which project the engine has is said to be the **active project**, and the engine is then active for that project.

Let's call the currently visible project **current project**. The user may change the current project by switching between them using tabs at the top, but that does not change the active project.

If the user wants to change the active project to the current project, they will need to do that explicitly.

Reply:
Okay, that fits my understanding. I'll leave this unresolved until I add project switching and then poke you about reviewing it.

Reply:
I think I'll leave this for a later PR. I need to refine engine lifecycle management, as well as add the necessary interface elements, and I want to do that all at once. For now I'll just assume the engine is always running, because for now it is.